File: ImasLiveDB-Android/app/src/main/kotlin/com/fugaif/imaslivedb/data/model/Song.kt
package com.fugaif.imaslivedb.data.model

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey

@Entity(
    tableName = "songs",
    indices = [
        Index(name = "idx_songs_brand", value = ["brand_id"]),
        Index(name = "idx_songs_composer", value = ["composer"])
    ]
)
data class Song(
    @PrimaryKey
    @ColumnInfo(name = "id")
    val id: String,

    @ColumnInfo(name = "title")
    val title: String,

    @ColumnInfo(name = "title_kana")
    val titleKana: String?,

    @ColumnInfo(name = "brand_id")
    val brandId: String?,

    @ColumnInfo(name = "song_type")
    val songType: String,

    @ColumnInfo(name = "release_date")
    val releaseDate: String?,

    @ColumnInfo(name = "duration_sec")
    val durationSec: Int?,

    @ColumnInfo(name = "composer")
    val composer: String?,

    @ColumnInfo(name = "lyricist")
    val lyricist: String?,

    @ColumnInfo(name = "arranger")
    val arranger: String?,

    @ColumnInfo(name = "cd_series")
    val cdSeries: String?,

    @ColumnInfo(name = "cd_title")
    val cdTitle: String?,

    @ColumnInfo(name = "artwork_url")
    val artworkUrl: String?,

    @ColumnInfo(name = "preview_url")
    val previewUrl: String?,

    @ColumnInfo(name = "apple_music_id")
    val appleMusicId: String?,

    @ColumnInfo(name = "apple_music_album_id")
    val appleMusicAlbumId: String?,

    @ColumnInfo(name = "isrc")
    val isrc: String?,

    @ColumnInfo(name = "lyrics_url")
    val lyricsUrl: String?,

    @ColumnInfo(name = "parent_song_id")
    val parentSongId: String?,

    @ColumnInfo(name = "singer_label")
    val singerLabel: String?,

    @ColumnInfo(name = "unit_name")
    val unitName: String?,

    @ColumnInfo(name = "unit_id")
    val unitId: String?,

    @ColumnInfo(name = "series_group")
    val seriesGroup: String? = null,

    /**
     * この曲がどのユニットの版のものか ([UnitVersion.id])。null = 無印。
     *
     * ユニットは 1 行のままで、版の違いは曲側が指す。ユニット単位のフラグにすると
     * リブート前後の曲を区別できない。判定は [UnitVersion.code] で行うこと。
     */
    @ColumnInfo(name = "unit_version_id")
    val unitVersionId: String? = null,

    /**
     * 合同曲 (コラボ曲) で、brand_id 以外に参加しているブランド (カンマ区切り)。
     * events の joint_brand_ids と同じ形で、参加ブランド全部の曲一覧に出すために使う。
     * **在籍の重なりでは入れない** — ML の曲に 765AS の面々が居るのも、876 の曲に
     * 秋月涼が居るのも合同ではない。
     */
    @ColumnInfo(name = "joint_brand_ids")
    val jointBrandIds: String? = null,

    /**
     * シリーズ横断の合同曲か。判断は人が持つ (原唱者のブランドから導くと在籍の重なりを
     * 合同と取り違える)。立てるなら [jointBrandIds] も入れる。
     */
    @ColumnInfo(name = "is_collab")
    val isCollab: Boolean = false
) {
    val isRemix: Boolean get() = parentSongId != null

    /** 参加ブランド ([brandId] が先頭、続いて [jointBrandIds])。 */
    val brandIds: List<String>
        get() = (listOfNotNull(brandId) + (jointBrandIds?.split(",") ?: emptyList()))
            .filter { it.isNotEmpty() }
}

File: ImasLiveDB/Models/Song.swift
import Foundation
import GRDB

struct Song: Codable, FetchableRecord, PersistableRecord, Identifiable, Hashable, Sendable {
    static let databaseTableName = "songs"

    var id: String
    var title: String
    var titleKana: String?
    var brandId: String?
    var songType: String
    var releaseDate: String?
    var durationSec: Int?
    var composer: String?
    var lyricist: String?
    var arranger: String?
    var cdSeries: String?
    var cdTitle: String?
    var artworkUrl: String?
    var previewUrl: String?
    var appleMusicId: String?
    var appleMusicAlbumId: String?
    var isrc: String?
    var lyricsUrl: String?
    var parentSongId: String?
    var singerLabel: String?
    var unitName: String?
    var unitId: String?
    /// 上位シリーズ (例: LIVE THE@TER FORWARD)。DB には存在するが宣言漏れで
    /// 同期のたび NULL に戻っていたため CodingKeys に追加。
    var seriesGroup: String?
    /// この曲がどのユニットの版のものか (`UnitVersion.id`)。nil = 無印。
    ///
    /// ユニットは 1 行のままで、版の違いは曲側が指す。ユニット単位のフラグにすると
    /// リブート前後の曲を区別できない。判定は `UnitVersion.code` で行うこと。
    var unitVersionId: String?
    /// 合同曲 (コラボ曲) で、`brandId` 以外に参加しているブランド (カンマ区切り)。
    /// `Event.jointBrandIds` と同じ形で、参加ブランド全部の曲一覧に出すために使う。
    /// **在籍の重なりでは入れない** — ML の曲に 765AS の面々が居るのも、876 の曲に
    /// 秋月涼が居るのも合同ではない。
    var jointBrandIds: String?
    /// シリーズ横断の合同曲か。判断は人が持つ (原唱者のブランドから導くと在籍の重なりを
    /// 合同と取り違える)。立てるなら `jointBrandIds` も入れる。
    var isCollab: Bool = false

    enum CodingKeys: String, CodingKey {
        case id, title, composer, lyricist, arranger, isrc
        case titleKana = "title_kana"
        case brandId = "brand_id"
        case songType = "song_type"
        case releaseDate = "release_date"
        case durationSec = "duration_sec"
        case cdSeries = "cd_series"
        case cdTitle = "cd_title"
        case artworkUrl = "artwork_url"
        case previewUrl = "preview_url"
        case appleMusicId = "apple_music_id"
        case appleMusicAlbumId = "apple_music_album_id"
        case lyricsUrl = "lyrics_url"
        case parentSongId = "parent_song_id"
        case singerLabel = "singer_label"
        case unitName = "unit_name"
        case unitId = "unit_id"
        case seriesGroup = "series_group"
        case unitVersionId = "unit_version_id"
        case jointBrandIds = "joint_brand_ids"
        case isCollab = "is_collab"
    }

    var isRemix: Bool { parentSongId != nil }

    /// 参加ブランド (`brandId` が先頭、続いて `jointBrandIds`)。
    var brandIds: [String] {
        ([brandId].compactMap { $0 } + (jointBrandIds?.split(separator: ",").map(String.init) ?? []))
            .filter { !$0.isEmpty }
    }

    /// 日本語表示用の楽曲タイプラベル
    var songTypeLabel: String {
        switch songType {
        case "solo": return "ソロ"
        // "group" は廃止したが、古い local DB が CloudKit pull されるまでの互換ラベル
        case "unit", "group": return "ユニット"
        case "all": return "全体曲"
        case "original": return "オリジナル"
        case "unknown": return "不明"
        default: return songType
        }
    }

    // MARK: - Associations

    static let brand = belongsTo(Brand.self)
    static let songArtists = hasMany(SongArtist.self)
    static let artists = hasMany(Idol.self, through: songArtists, using: SongArtist.idol)
    static let setlistItems = hasMany(SetlistItem.self)

    var brand: QueryInterfaceRequest<Brand> { request(for: Song.brand) }
    var artists: QueryInterfaceRequest<Idol> { request(for: Song.artists) }
    var setlistItems: QueryInterfaceRequest<SetlistItem> { request(for: Song.setlistItems) }
}
